File: frontend/src/lib/markdownTaskList.ts
const TASK_ITEM_RE = /^(\s*)([-*+])\s+\[([ xX])\]\s+(.*)$/;

export interface TaskListItem {
  lineIndex: number;
  checked: boolean;
  text: string;
}

export type MarkdownBlock =
  | { type: "html"; content: string }
  | { type: "taskList"; items: TaskListItem[] };

export function hasTaskItems(markdown: string): boolean {
  return markdown.split("\n").some((line) => TASK_ITEM_RE.test(line));
}

export function parseMarkdownBlocks(markdown: string): MarkdownBlock[] {
  const lines = markdown.split("\n");
  const blocks: MarkdownBlock[] = [];
  let htmlLines: string[] = [];
  let taskItems: TaskListItem[] = [];

  const flushHtml = (): void => {
    if (htmlLines.length > 0) {
      blocks.push({ type: "html", content: htmlLines.join("\n") });
      htmlLines = [];
    }
  };

  const flushTaskList = (): void => {
    if (taskItems.length > 0) {
      blocks.push({ type: "taskList", items: [...taskItems] });
      taskItems = [];
    }
  };

  for (let index = 0; index < lines.length; index += 1) {
    const line = lines[index] ?? "";
    const match = TASK_ITEM_RE.exec(line);
    if (match) {
      flushHtml();
      taskItems.push({
        lineIndex: index,
        checked: match[3]?.toLowerCase() === "x",
        text: match[4] ?? "",
      });
      continue;
    }
    if (line.trim() === "" && taskItems.length > 0) {
      flushTaskList();
      htmlLines.push(line);
      continue;
    }
    flushTaskList();
    htmlLines.push(line);
  }

  flushHtml();
  flushTaskList();
  return blocks;
}

export function toggleTaskItemLocal(markdown: string, lineIndex: number): string {
  const lines = markdown.split("\n");
  const line = lines[lineIndex];
  if (line === undefined) {
    throw new Error(`Invalid line_index: ${lineIndex}`);
  }
  const match = TASK_ITEM_RE.exec(line);
  if (!match) {
    throw new Error(`Line ${lineIndex} is not a task list item`);
  }
  const indent = match[1] ?? "";
  const bullet = match[2] ?? "-";
  const check = match[3] ?? " ";
  const rest = match[4] ?? "";
  const newCheck = check.toLowerCase() === "x" ? " " : "x";
  lines[lineIndex] = `${indent}${bullet} [${newCheck}] ${rest}`;
  return lines.join("\n");
}

File: backend/app/services/markdown_task_list.py
"""GFM task list helpers for dashboard markdown text widgets."""

import re

TASK_ITEM_RE = re.compile(r"^(\s*)([-*+])\s+\[([ xX])\]\s+(.*)$")


def has_task_items(markdown: str) -> bool:
    """Return True when markdown contains at least one GFM task list line."""
    return any(TASK_ITEM_RE.match(line) for line in markdown.splitlines())


def parse_task_line_indices(markdown: str) -> list[int]:
    """Return 0-based line indices of GFM task list items in the markdown."""
    return [i for i, line in enumerate(markdown.splitlines()) if TASK_ITEM_RE.match(line)]


def toggle_task_item(markdown: str, line_index: int) -> str:
    """Flip `[ ]` ↔ `[x]` on the task list line at ``line_index``.

    Raises:
        ValueError: When the line index is out of range or not a task item.
    """
    lines = markdown.split("\n")
    if line_index < 0 or line_index >= len(lines):
        raise ValueError(f"Invalid line_index: {line_index}")
    match = TASK_ITEM_RE.match(lines[line_index])
    if match is None:
        raise ValueError(f"Line {line_index} is not a task list item")
    indent, bullet, check, rest = match.groups()
    new_check = " " if check.lower() == "x" else "x"
    lines[line_index] = f"{indent}{bullet} [{new_check}] {rest}"
    return "\n".join(lines)


def update_task_item(markdown: str, line_index: int, text: str) -> str:
    """Replace the label text on a task list line, preserving checked state.

    Raises:
        ValueError: When the line index is out of range or not a task item.
    """
    lines = markdown.split("\n")
    if line_index < 0 or line_index >= len(lines):
        raise ValueError(f"Invalid line_index: {line_index}")
    match = TASK_ITEM_RE.match(lines[line_index])
    if match is None:
        raise ValueError(f"Line {line_index} is not a task list item")
    indent, bullet, check, _rest = match.groups()
    lines[line_index] = f"{indent}{bullet} [{check}] {text}"
    return "\n".join(lines)


def remove_task_item(markdown: str, line_index: int) -> str:
    """Remove a task list line from markdown.

    Raises:
        ValueError: When the line index is out of range or not a task item.
    """
    lines = markdown.split("\n")
    if line_index < 0 or line_index >= len(lines):
        raise ValueError(f"Invalid line_index: {line_index}")
    if TASK_ITEM_RE.match(lines[line_index]) is None:
        raise ValueError(f"Line {line_index} is not a task list item")
    del lines[line_index]
    if (
        line_index < len(lines)
        and lines[line_index].strip() == ""
        and line_index > 0
        and lines[line_index - 1].strip() == ""
    ):
        del lines[line_index]
    return "\n".join(lines)


def update_or_remove_task_item(markdown: str, line_index: int, text: str) -> str:
    """Update task item label text, or remove the line when text is blank."""
    if text.strip() == "":
        return remove_task_item(markdown, line_index)
    return update_task_item(markdown, line_index, text)
